File: 1043.java
import java.util.*;
import java.io.*;
public class Main {
	
	static StreamTokenizer in = new StreamTokenizer(new BufferedReader(new InputStreamReader(System.in)));
	static PrintWriter out = new PrintWriter(new OutputStreamWriter(System.out));
	
	static Scanner input = new Scanner(System.in);
	static Queue<Integer> queue = new LinkedList<Integer>();
	static boolean flag = false, complete = true;
	static int n, node, now;
	static int[][] tree = new int[1000001][3];
 
	public static void main(String[] args) throws IOException {
		
		in.nextToken();
		n = (int) in.nval;
//		n = input.nextInt();
		for (int i = 1; i < n; i++) {
			in.nextToken();
			node = (int) in.nval;
//			node = input.nextInt();
			tree[i][0] = node;
			if (tree[node][1] == 0) tree[node][1] = i;
			else tree[node][2] = i;
		}
		if (tree[0][1] != 0) queue.offer(tree[0][1]);
		if (tree[0][2] != 0) queue.offer(tree[0][2]);
 
		while (!queue.isEmpty()) {
			now = queue.poll();
			if ((tree[now][1] == 0)&&(tree[now][2] == 0)) flag = true;
			if (tree[now][1] != 0) {
				queue.offer(tree[now][1]);
				if (flag ==true) {
					complete = false;
					break;
				}
			}
			if (tree[now][2] != 0) queue.offer(tree[now][2]);
		}
		System.out.println(complete);
	}
}

File: 1043.cpp
#include <iostream>
#include <queue>
 
using namespace std;
struct treedata{
    int l, r, pre;
} tree[1000001];
int n, i, node, now;
bool flag,complete = true;
queue<int> que;
 
int main(){
    cin>>n;
    for (int i = 1; i < n; i++){
        cin>>node;
        tree[i].pre = node;
        if (tree[node].l == 0) tree[node].l = i;
        else tree[node].r = i;
    }
    if (tree[0].l) que.push(tree[0].l);
    if (tree[0].r) que.push(tree[0].r);
    while (!que.empty()){
        now = que.front();
        que.pop();
        if (!tree[now].l&&!tree[now].r) flag = true;
        if (tree[now].l){
            que.push(tree[now].l);
            if (flag){
                complete = false;
                break;
            }
        }
        if (tree[now].r) que.push(tree[now].r);
    }
    if (complete) cout<<"true"; else cout<<"false";
    return 0;
}